src/TwitchLib.Communication/Services/ConnectionWatchDog.cs

```diff
@@ -0,0 +1,139 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.Extensions.Logging;
+
+using TwitchLib.Communication.Clients;
+using TwitchLib.Communication.Events;
+using TwitchLib.Communication.Extensions;
+
+namespace TwitchLib.Communication.Services
+{
+
+    /// <summary>
+    ///     just to check connection state
+    /// </summary>
+    internal class ConnectionWatchDog<T> where T : IDisposable
+    {
+
+        #region properties private
+        private ILogger? LOGGER { get; }
+        private AClientBase<T> Client { get; }
+        /// <summary>
+        ///     <list>
+        ///         <item>
+        ///             should only be set to a new instance in <see cref="StartMonitorTask()"/>
+        ///         </item>
+        ///         <item>
+        ///             should only be set to <see langword="null"/> in <see cref="Stop()"/>
+        ///         </item>
+        ///     </list>
+        /// </summary>
+        private CancellationTokenSource CancellationTokenSource { get; set; }
+        private CancellationToken Token => CancellationTokenSource.Token;
+        private int MonitorTaskDelayInMilliseconds => 200;
+        #endregion properties private
+
+
+        #region ctors
+        internal ConnectionWatchDog(AClientBase<T> client,
+                                    ILogger? logger = null)
+        {
+            LOGGER = logger;
+            Client = client;
+        }
+        #endregion ctors
+
+
+        #region methods internal
+
+        internal Task StartMonitorTask()
+        {
+            LOGGER?.TraceMethodCall(GetType());
+            // we dont want to start more than one WatchDog
+            if (CancellationTokenSource != null)
+            {
+                Exception ex = new InvalidOperationException("Monitor Task cant be started more than once!");
+                LOGGER?.LogExceptionAsError(GetType(), ex);
+                throw ex;
+            }
+            // this should be the only place where a new instance of CancellationTokenSource is set
+            CancellationTokenSource = new CancellationTokenSource();
+            if (Token == null)
+            {
+                Exception ex = new InvalidOperationException($"{nameof(Token)} was null!");
+                LOGGER?.LogExceptionAsError(GetType(), ex);
+                Client.RaiseFatal(ex);
+                throw ex;
+            }
+            return Task.Run(MonitorTaskAction, Token);
+        }
+
+        internal void Stop()
+        {
+            LOGGER?.TraceMethodCall(GetType());
+            CancellationTokenSource?.Cancel();
+            // give MonitorTaskAction a chance to catch cancellation
+            // otherwise it may result in an Exception
+            Task.Delay(MonitorTaskDelayInMilliseconds * 2).GetAwaiter().GetResult();
+            CancellationTokenSource?.Dispose();
+            // set it to null for the check within this.StartMonitorTask()
+            CancellationTokenSource = null;
```

Comment:
Warning: Cannot convert null literal to non-nullable reference type. See my previous comment.